File: packages/python/src/daytona_sdk/git.py
from typing import List, Optional, TYPE_CHECKING
from api_client import (
    GitStatus,
    ListBranchResponse,
    Match,
    Workspace as WorkspaceInstance,
    WorkspaceToolboxApi,
)

if TYPE_CHECKING:
    from .workspace import Workspace


class Git:
    def __init__(
        self,
        workspace: "Workspace",
        toolbox_api: WorkspaceToolboxApi,
        instance: WorkspaceInstance,
    ):
        self.workspace = workspace
        self.toolbox_api = toolbox_api
        self.instance = instance

    def add(self, path: str, files: List[str]) -> None:
        self.toolbox_api.git_add_files(
            workspace_id=self.instance.id,
            project_id="main",  # todo: remove this after project refactor
            params={"path": path, "files": files},
        )

    def branches(self, path: str) -> ListBranchResponse:
        return self.toolbox_api.git_branch_list(
            workspace_id=self.instance.id,
            project_id="main",  # todo: remove this after project refactor
            path=path,
        )

    def clone(
        self,
        url: str,
        path: str,
        branch: Optional[str] = None,
        commit_id: Optional[str] = None,
        username: Optional[str] = None,
        password: Optional[str] = None,
    ) -> None:
        self.toolbox_api.git_clone_repository(
            workspace_id=self.instance.id,
            project_id="main",  # todo: remove this after project refactor
            params={
                "url": url,
                "branch": branch,
                "path": path,
                "username": username,
                "password": password,
                "commitId": commit_id,
            },
        )

    def commit(self, path: str, message: str, author: str, email: str) -> None:
        self.toolbox_api.git_commit_changes(
            workspace_id=self.instance.id,
            project_id="main",  # todo: remove this after project refactor
            params={"path": path, "message": message, "author": author, "email": email},
        )

    def push(
        self, path: str, username: Optional[str] = None, password: Optional[str] = None
    ) -> None:
        self.toolbox_api.git_push_changes(
            workspace_id=self.instance.id,
            project_id="main",  # todo: remove this after project refactor
            params={"path": path, "username": username, "password": password},
        )

    def pull(
        self, path: str, username: Optional[str] = None, password: Optional[str] = None
    ) -> None:
        self.toolbox_api.git_pull_changes(
            workspace_id=self.instance.id,
            project_id="main",  # todo: remove this after project refactor
            params={"path": path, "username": username, "password": password},
        )

    def status(self, path: str) -> GitStatus:
        return self.toolbox_api.git_git_status(
            workspace_id=self.instance.id,
            project_id="main",  # todo: remove this after project refactor
            path=path,
        )

File: packages/typescript/src/Git.ts
import {
  GitStatus,
  ListBranchResponse,
  Match,
  Workspace as WorkspaceInstance,
  WorkspaceToolboxApi,
} from './client'
import { Workspace } from './Workspace'

export class Git {
  constructor(
    private readonly workspace: Workspace,
    private readonly toolboxApi: WorkspaceToolboxApi,
    private readonly instance: WorkspaceInstance,
  ) {}
  public async add(path: string, files: string[]): Promise<void> {
    await this.toolboxApi.gitAddFiles({
      workspaceId: this.instance.id,
      projectId: 'main', //  todo: remove this after project refactor
      params: {
        path,
        files,
      },
    })
  }

  public async branches(path: string): Promise<ListBranchResponse> {
    return await this.toolboxApi.gitBranchList({
      workspaceId: this.instance.id,
      projectId: 'main', //  todo: remove this after project refactor
      path,
    })
  }

  public async clone(
    url: string,
    path: string,
    branch?: string,
    commitId?: string,
    username?: string,
    password?: string,
  ): Promise<void> {
    await this.toolboxApi.gitCloneRepository({
      workspaceId: this.instance.id,
      projectId: 'main', //  todo: remove this after project refactor
      params: {
        url: url,
        branch: branch,
        path,
        username,
        password,
        commitId,
      },
    })
  }

  public async commit(
    path: string,
    message: string,
    author: string,
    email: string,
  ): Promise<void> {
    await this.toolboxApi.gitCommitChanges({
      workspaceId: this.instance.id,
      projectId: 'main', //  todo: remove this after project refactor
      params: {
        path,
        message,
        author,
        email,
      },
    })
  }

  public async push(
    path: string,
    username?: string,
    password?: string,
  ): Promise<void> {
    await this.toolboxApi.gitPushChanges({
      workspaceId: this.instance.id,
      projectId: 'main', //  todo: remove this after project refactor
      params: {
        path,
        username,
        password,
      },
    })
  }

  public async pull(
    path: string,
    username?: string,
    password?: string,
  ): Promise<void> {
    await this.toolboxApi.gitPullChanges({
      workspaceId: this.instance.id,
      projectId: 'main', //  todo: remove this after project refactor
      params: {
        path,
        username,
        password,
      },
    })
  }

  public async status(path: string): Promise<GitStatus> {
    return await this.toolboxApi.gitGitStatus({
      workspaceId: this.instance.id,
      projectId: 'main', //  todo: remove this after project refactor
      path,
    })
  }
}
